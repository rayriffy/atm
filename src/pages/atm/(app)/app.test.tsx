import { render, act } from '@testing-library/react'
import { PropsWithChildren } from 'react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import Page from './app'

const propsSpy = vi.fn()
const {
  handlePadSpy,
  useKeyboardInputSpy,
  PadSpy,
  BalanceSpy,
  ErrorDecoderSpy,
  InputSpy,
  InventorySpy,
  OverdrawnSpy,
  WithdrawButtonSpy,
} = vi.hoisted(() => ({
  handlePadSpy: vi.fn(),
  useKeyboardInputSpy: vi.fn(),
  PadSpy: vi.fn((props: PropsWithChildren) => {
    propsSpy(props)
    return <div>Pad{props.children}</div>
  }),
  BalanceSpy: vi.fn(() => <div>Balance</div>),
  ErrorDecoderSpy: vi.fn(() => <div>ErrorDecoder</div>),
  InputSpy: vi.fn(() => <div>Input</div>),
  InventorySpy: vi.fn(() => <div>Inventory</div>),
  OverdrawnSpy: vi.fn(() => <div>Overdrawn</div>),
  WithdrawButtonSpy: vi.fn(() => <div>WithdrawButton</div>),
}))

vi.mock('$pad', () => ({
  Pad: PadSpy,
}))

vi.mock('$user/handlePad', () => ({
  handlePad: handlePadSpy,
}))

vi.mock('$user/useKeyboardInput', () => ({
  useKeyboardInput: useKeyboardInputSpy,
}))

vi.mock('$user/balance', () => ({
  Balance: BalanceSpy,
}))

vi.mock('$user/errorDecoder', () => ({
  ErrorDecoder: ErrorDecoderSpy,
}))

vi.mock('$user/input', () => ({
  Input: InputSpy,
}))

vi.mock('$user/inventory', () => ({
  Inventory: InventorySpy,
}))

vi.mock('$user/overdrawn', () => ({
  Overdrawn: OverdrawnSpy,
}))

vi.mock('$user/withdrawButton', () => ({
  WithdrawButton: WithdrawButtonSpy,
}))

describe('<Page />', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('should render page with all components', () => {
    const { container } = render(<Page />)

    expect(PadSpy).toBeCalledTimes(1)
    expect(BalanceSpy).toBeCalledTimes(1)
    expect(ErrorDecoderSpy).toBeCalledTimes(1)
    expect(InputSpy).toBeCalledTimes(1)
    expect(InventorySpy).toBeCalledTimes(1)
    expect(OverdrawnSpy).toBeCalledTimes(1)
    expect(WithdrawButtonSpy).toBeCalledTimes(1)

    expect(container).toMatchSnapshot()
  })

  it('should pass onClick to Pad', () => {
    render(<Page />)

    expect(PadSpy).toBeCalledTimes(1)

    act(() => {
      propsSpy.mock.calls[0][0].onClick()
    })

    expect(handlePadSpy).toBeCalledTimes(1)
  })
})
