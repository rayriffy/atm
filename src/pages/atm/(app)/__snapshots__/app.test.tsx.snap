// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Page /> > should render page with all components 1`] = `
<div>
  <div
    class="chakra-stack css-kmvjml"
  >
    <div
      class="chakra-stack css-1igwmid"
    >
      <h2
        class="chakra-heading css-0"
      >
        Withdraw
      </h2>
      <div
        class="css-17xejub"
      />
      <div>
        Balance
      </div>
    </div>
    <div>
      Input
    </div>
    <div>
      ErrorDecoder
    </div>
  </div>
  <div>
    Pad
    <div>
      Inventory
    </div>
    <div
      class="chakra-stack css-1igwmid"
    >
      <div>
        Overdrawn
      </div>
      <div
        class="css-17xejub"
      />
      <div>
        WithdrawButton
      </div>
    </div>
  </div>
</div>
`;
